### PR TITLE
Automatic duckpan app restart on IA update

### PR DESCRIPTION
**Pull request by zachthompson**
**Which issues (if any) does this change solve? (please link them here)**
https://github.com/duckduckgo/p5-app-duckpan/issues/52

**Briefly describe your fix. Are there any important details we should know about?**
Alternate strategy to https://github.com/duckduckgo/p5-app-duckpan/pull/175 code reloading.  Adds the role App::DuckPAN::Restart which _restarts_ the running duckpan application automatically when instant answer code is changed.

Incorporates https://github.com/duckduckgo/p5-app-duckpan/pull/190 as well since the current query console does not exit cleanly when killed.

### CODE REVIEW

File: lib/App/DuckPAN/Restart.pm [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    # will keep (re)starting the server until user ctrl-c
+    while(1){
+        defined(my $app = fork) or die 'Failed to fork application';
+        unless($app){ # app kid
+            $self->_run_app($args);
+            exit 0;
+        }
+
+        my $fmon = fork;
```
Comment (nilnilnil) [inline lib/App/DuckPAN/Restart.pm]: why the different checking strategies?

**Comment by nilnilnil** [comment]
:+1:

**Comment by nilnilnil** [comment]
Failing tests though.

File: lib/App/DuckPAN/Restart.pm [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    my ($self, $args) = @_;
+
+    # will keep (re)starting the server until user ctrl-c
+    while(1){
+        defined(my $app = fork) or die 'Failed to fork application';
+        unless($app){ # app kid
+            $self->_run_app($args);
+            exit 0;
+        }
+
```
Comment (nilnilnil) [inline lib/App/DuckPAN/Restart.pm]: Comment why you decided to do a different fork & check strategy.

File: lib/App/DuckPAN/Restart.pm [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+        unless($fmon){ # file monitor kid
+            unless(defined $fmon){
+                kill SIGTERM => $app;
+                die 'Failed to fork file monitor';
+            }
+            $self->_monitor_directories;
+            exit 0;
+        }
+
+        # wait for one them to exit
```
Comment (nilnilnil) [inline lib/App/DuckPAN/Restart.pm]: comment that -1 means all child processes

File: lib/App/DuckPAN/Restart.pm [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    }
+}
+
+# Monitors development directories for file changes.  Tries to get the
+# list of directories in a general way.  This subroutine 
+# blocks, so when it returns we know there's been a change
+sub _monitor_directories {
+    my $self  = shift;
+
+    # Find all of the directories that neeed to monitored
```
Comment (nilnilnil) [inline lib/App/DuckPAN/Restart.pm]: add an issue or a comment here that we should improve this / make a function for it.

File: lib/App/DuckPAN/Restart.pm (hunk 2 above)
Comment (zachthompson) [inline lib/App/DuckPAN/Restart.pm]: Exactly.  We need to clean up the spawned app before exiting with the second fork.  If we want the structure to be the same, we can change the first fork to the latter's.

File: lib/App/DuckPAN/Restart.pm (hunk 4 above)
Comment (zachthompson) [inline lib/App/DuckPAN/Restart.pm]: https://github.com/duckduckgo/p5-app-duckpan/issues/200

Comment (nilnilnil) [inline lib/App/DuckPAN/Restart.pm]: :+1:

**Comment by nilnilnil** [comment]
@zachthompson still needs merge conflict resolution

File: lib/App/DuckPAN/Restart.pm [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            unless(defined $fmon){
+                kill SIGTERM => $app;
+                die 'Failed to fork file monitor';
+            }
+            $self->_monitor_directories;
+            exit 0;
+        }
+
+        # wait for one them to exit. -1 waits for all children
+        my $pid = waitpid -1, 0;
```
Comment (zachthompson) [inline lib/App/DuckPAN/Restart.pm]: I think there might be a slight race condition here when *outside of an IA directory* if the pid for the fmon is reaped first.  Let me have a look at it.  I test this very condition this morning and it was fine but running it this afternoon for something unrelated triggered it.

**Comment by russellholt** [comment]
@zachthompson let me know when you're ready.

**Comment by nilnilnil** [comment]
@zachthompson are you all set here?

**Comment by zachthompson** [comment]
@nilnilnil @russellholt ready

**Comment by nilnilnil** [comment]
:+1:

File: lib/App/DuckPAN/Restart.pm [hunk 6]
```diff
@@ -0,0 +1,107 @@
+package App::DuckPAN::Restart;
+# ABSTRACT: Automatic restarting of application on file change
+
+use File::Find::Rule;
+use Filesys::Notify::Simple;
```
Comment (moollaza) [inline lib/App/DuckPAN/Restart.pm]: @zachthompson These need to be added to the dist.ini please :)

**Pull request merged**